File: src/main/java/au/id/micolous/metrodroid/transit/en1545/En1545FixedHex.java
/*
 * En1545Fixed.java
 *
 * Copyright 2018 Google
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package au.id.micolous.metrodroid.transit.en1545;

import java.util.Locale;

public class En1545FixedHex implements En1545Field {
    private final int mLen;
    private final String mName;

    public En1545FixedHex(String name, int len) {
        mName = name;
        mLen = len;
    }

    @SuppressWarnings("AssignmentToForLoopParameter")
    @Override
    public int parseField(byte[] b, int off, String path, En1545Parsed holder, En1545Bits bitParser) {
        StringBuilder res = new StringBuilder();
        try {
            for (int i = mLen; i > 0; ) {
                if (i >= 8) {
                    res.insert(0, String.format(Locale.ENGLISH, "%02x",
                            bitParser.getBitsFromBuffer(b, off + i - 8, 8)));
                    i -= 8;
                    continue;
                }
                if (i >= 4){
                    res.insert(0, String.format(Locale.ENGLISH, "%01x",
                            bitParser.getBitsFromBuffer(b, off + i - 4, 4)));
                    i -= 4;
                    continue;
                }
                res.insert(0, String.format(Locale.ENGLISH, "%x",
                        bitParser.getBitsFromBuffer(b, off, i)));
                break;
            }
            holder.insertString(mName, path, res.toString());
        } catch (Exception e) {
        }
        return off + mLen;
    }
}

File: src/commonMain/kotlin/au/id/micolous/metrodroid/transit/en1545/En1545FixedHex.kt
/*
 * En1545Fixed.kt
 *
 * Copyright 2018 Google
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package au.id.micolous.metrodroid.transit.en1545

import au.id.micolous.metrodroid.multi.Log

import au.id.micolous.metrodroid.util.ImmutableByteArray

class En1545FixedHex(private val mName: String, private val mLen: Int) : En1545Field {

    override fun parseField(b: ImmutableByteArray, off: Int, path: String, holder: En1545Parsed, bitParser: En1545Bits): Int {
        var res = ""
        try {
            var i = mLen
            while (i > 0) {
                if (i >= 8) {
                    var t = bitParser(b, off + i - 8, 8).toString(16)
                    if (t.length == 1)
                        t = "0$t"
                    res = t + res
                    i -= 8
                    continue
                }
                if (i >= 4) {
                    res = bitParser(b, off + i - 4, 4).toString(16) + res
                    i -= 4
                    continue
                }
                res = bitParser(b, off, i).toString(16) + res
                break
            }
            holder.insertString(mName, path, res)
        } catch (e: Exception) {
            Log.w("En1545FixedHex", "Overflow when parsing en1545", e)
        }

        return off + mLen
    }
}
